tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/README.md

```diff
@@ -0,0 +1,402 @@
+├── src                         # React app source code
+│   ├── components
+│   │   ├── App.tsx             # Main React component
+│   ├── core
+│   │   ├── blockchain.ts       # TS module responsible for interaction with blockchain and contracts
+│   ├── config/                 # Folder with contract ABIs and deployed contract addresses
+│   └── ...
+├── test                        # Contract tests
+└── ...
+```
+
+### Contracts
+
+#### ExampleNFT.sol
+
+`ExampleNFT` is a simple ERC721 contract with automated sequential token id assignment
+
+```js
+function mint() external {
+    _mint(msg.sender, nextTokenId);
+    nextTokenId++;
+}
+```
+
+This contract extends `ERC721Enumerable` implementation created by the `@openzeppelin` team, which adds view functions for listing all tokens and tokens owned by a user.
+
+#### Marketplace.sol
+
+`Marketplace` is an NFT marketplace contract, which allows to post sell orders for any NFT token that follows [EIP-721 non-fungible token standard](https://eips.ethereum.org/EIPS/eip-721). It has the following functions:
+
+```js
+
+// Created a new sell order
+// This function requires approval for transfer on the specified NFT token
+function postSellOrder(address nftContractAddress, uint256 tokenId, uint256 price) external {}
+
+// Only order creator can call this function
+function cancelOrder(uint256 orderId) external {}
+
+// Allows to get info about all orders (including canceled, and executed ones)
+function getAllOrders() public view returns (SellOrder[] memory) {}
+
+// Returns expected price in ETH for the given order
+function getPrice(uint256 orderId) public view returns (uint256) {}
+
+// Requires sending at least the minimal amount of ETH
+function buy(uint256 orderId) external payable {}
+
+```
+
+The implementation is quite straightforward, so we won't describe it here. You can check the full contract code in the [contracts/Marketplace.sol.](contracts/Marketplace.sol)
+
+#### StableMarketplace.sol
+
+`StableMarketplace` is the marketplace contract with the stable price support. It extends the `Marketplace.sol` implementation and only overrides its `_getPriceFromOrder` function.
+This contract will integrate RedStone oracles functionalities and will be described later.
+
+### Frontend
+
+You can check the code of the React app in the `src` folder. We tried to simplify it as much as possible and leave only the core marketplace functions.
+
+The main UI logic is located in the `App.tsx` file, and the contract interaction logic is in the `blockchain.ts` file.
+
+If you take a look into the `blockchain.ts` file code, you'll notice that each contract call that needs to process RedStone data is made on a contract instance, that was wrapped by [@redstone-finance/evm-connector](https://www.npmjs.com/package/@redstone-finance/evm-connector).
+
+### Tests
+
+We've used hardhat test framework to contract tests. All the tests are located in the [test](test/) folder.
```

Comment:
Do the tests all pass on your end? When I run the tests I get:

```bash
yarn test
yarn run v1.22.19
$ NODE_ENV=test hardhat test --network zkSyncTestnet


  ExampleNFT
    ✔ Should deploy ExampleNFT (2136ms)
    ✔ Should mint 2 NFTs (1778ms)
    ✔ User1 should approve spending and User2 should spend (2838ms)

  Marketplace core functions test
    ✔ Should deploy contracts (3038ms)
    ✔ Should mint NFT (889ms)
    ✔ Seller should post sell order for token 1 with ETH price (2859ms)
    ✔ Should get all orders
Expected ETH amount: 1.0
    ✔ Buying should fail with smaller amount then seller requested (316ms)
Expected ETH amount: 1.0
    ✔ Buyer should buy token 1 for ETH price (917ms)
    ✔ Should post and cancel order (4811ms)

  Marketplace core functions test
    ✔ Should deploy contracts (2953ms)
    ✔ Should mint NFT (1901ms)
    ✔ Seller should post sell order for token 2 with stable USD price (2778ms)
    ✔ Should wrap marketplace contract with redstone wrapper
Expected ETH amount: 0.0
    1) Buying should fail with smaller amount then seller requested
Expected ETH amount: 0.0
    2) Buyer should buy token for USD price expressed in ETH


  14 passing (31s)
  2 failing

  1) Marketplace core functions test
       Buying should fail with smaller amount then seller requested:
     AssertionError: Expected transaction to be reverted
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Context.<anonymous> (test/stable-marketplace.test.ts:88:5)

  2) Marketplace core functions test
       Buyer should buy token for USD price expressed in ETH:
     Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={"reason":"execution reverted","code":"UNPREDICTABLE_GAS_LIMIT","method":"estimateGas","transaction":{"from":"0xa61464658AfeAf65CccaaFD3a512b69A83B77618","gasPrice":{"type":"BigNumber","hex":"0x0ee6b280"},"to":"0x0a67078A35745947A37A552174aFe724D8180c25","value":{"type":"BigNumber","hex":"0x00"},"data":"0xd96a094a0000000000000000000000000000000000000000000000000000000000000000455448000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025171e97dd018a845a3920000000200000010184df94c1cdf3dacc79d3081d78a5798364ee89ab96a76a9145e796ad9db325299c10dba57d6b4cb294ea811069d44477a21c718d6785174d52fa86bd6ef94a1b00013136393434333736373338343623302e322e352372656473746f6e652d6d61696e2d64656d6f000026000002ed57011e0000","type":0,"accessList":null},"error":{"reason":"processing response error","code":"SERVER_ERROR","body":"{\"jsonrpc\":\"2.0\",\"error\":{\"code\":3,\"message\":\"execution reverted\",\"data\":\"0x\"},\"id\":68}","error":{"code":3,"data":"0x"},"requestBody":"{\"method\":\"eth_estimateGas\",\"params\":[{\"gasPrice\":\"0xee6b280\",\"type\":\"0x0\",\"value\":\"0x0\",\"from\":\"0xa61464658afeaf65cccaafd3a512b69a83b77618\",\"to\":\"0x0a67078a35745947a37a552174afe724d8180c25\",\"data\":\"0xd96a094a0000000000000000000000000000000000000000000000000000000000000000455448000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025171e97dd018a845a3920000000200000010184df94c1cdf3dacc79d3081d78a5798364ee89ab96a76a9145e796ad9db325299c10dba57d6b4cb294ea811069d44477a21c718d6785174d52fa86bd6ef94a1b00013136393434333736373338343623302e322e352372656473746f6e652d6d61696e2d64656d6f000026000002ed57011e0000\"}],\"id\":68,\"jsonrpc\":\"2.0\"}","requestMethod":"POST","url":"http://localhost:3050"}}, tx={"data":"0xd96a094a0000000000000000000000000000000000000000000000000000000000000000455448000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025171e97dd018a845a3920000000200000010184df94c1cdf3dacc79d3081d78a5798364ee89ab96a76a9145e796ad9db325299c10dba57d6b4cb294ea811069d44477a21c718d6785174d52fa86bd6ef94a1b00013136393434333736373338343623302e322e352372656473746f6e652d6d61696e2d64656d6f000026000002ed57011e0000","to":{},"from":"0xa61464658AfeAf65CccaaFD3a512b69A83B77618","value":{"type":"BigNumber","hex":"0x00"},"type":0,"gasPrice":{},"nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
      at /Users/dennis/Documents/Code/tutorials/tutorials/zkSync-RedStone-stable-price-marketplace-tutorial/node_modules/@ethersproject/abstract-signer/src.ts/index.ts:301:31
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async Promise.all (index 7)



error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Reply:
Yes, but they will pass when you populate functions indicated in the tutorial like `buy` in dApp or `_getPriceFromOrder` in `StableMarketplace` contract